### ISSUE TITLE
There should be several current themes

### ISSUE THREAD

**Issue by CSurieux** [OrchardCMS/Orchard2#141]
May be switchable by placement file, content type, path or current layer.

**Comment by bleroy** [comment]
Why don't you implement IThemeSelector?

**Comment by CSurieux** [comment]
[Flags]
enum blnswrs long {
 slnce = 0,
 wht=1,
whch=2,
whr=4,
wh=8,
nt=16
is =32,
ths =64,
prcss =128,
in = 256,
prjct = 512,
th=1024,
sn't=2048,
it=4096,
myb=8192,
? = 16384,
}

string.Join(", ",((blnsrs)7408).ToString().split(new[]{','}).select(x => Rehydrat(x)).ForHuman(currentCulture)).ToList())

:)

**Comment by Jetski5822** [comment]
ITheneSelector is what you want as Bertrand mentioned.

**Comment by CSurieux** [comment]
Ha thanks, but this answer does not cover the idea of having placement file deciding ?

**Comment by Jetski5822** [comment]
Placement is about shape dispatching? It has nothing to do with Themes?

**Comment by CSurieux** [comment]
aren't times changing?

**Comment by CSurieux** [comment]
Look Nick, some information concerning display are in several places: placement.info, module.txt (layers), separate settings in site settings record containing xml, and some are not available because in code.
This is due to different strates of dev through O1 life.
May be it is time to decide some other organization, re-organize, of all this.
And the placement file seems ideal with its aggregate feature from modules.
And text files are the best place to be compatible with many different kind of env (Windows, mobile, Azure, other cloud, IOT, etc.)

**Comment by sfmskywalker** [comment]
The main responsibility of Placement.info is to determine where shapes created by content part drivers are to be placed in the Content shape. Mixing that concern with other concerns sounds wrong. Because a system (such as rendering) is implemented with sub systems (such as drivers, placement.info, theme selection, etc), that doesn't mean you can just merge these sub-systems. Well, suppose you could, but that would be a really bad implementation. Because, you know, SRP and such. Adding theme selection to placement.info would be mixing two separate concerns at the wrong level.

Better would be to implement a `IThemeSelector` that reads another text file with theme selection rules, like what you mentioned: based on content type, path or current layer, in case you want this to be configurable.

**Comment by CSurieux** [comment]
It was just an easy way to say: for this path place items here and use this theme, or for these content types place them here and use this theme.
The alternate offer of theme selector fragments the creation process between many tools, one of the small failures of O1. For a theme selector which says that for this path I want to use this theme, I need to create settings.
And I don't see the implication with SRP ???

**Comment by bleroy** [comment]
If you disagree with SRP, you disagree with almost the whole industry and with core principles in Orchard. I mean, if that is what you're disagreeing with, you are putting the bar extremely high for yourself. You would have to come up with an absolutely compelling scenario to justify going against a design principle that has such clear benefits.

You do however have an interesting scenario here, so I wouldn't be so quick to close the bug. It's just that the implementation you propose is not one we'd agree on.

What if we had a setting on each theme where you can specify a rule expression, exactly like we have rule expressions for activating layers? The difference would be that the result of the expression should be a priority rather than a Boolean. This way, each theme can decide to activate itself based on a rule that is decided by the site admin. That would be pretty sweet, in fact.

Also, there is no finite value of n for which 2^n = 0.

**Comment by CSurieux** [comment]
Ok not so bad :)

**Comment by sfmskywalker** [comment]
That's a great idea. Perhaps the theme rule would consist of two things: a Boolean expression exactly like widget layers (and elements), and a Priority field. That way all current functions can be reused. The theme selector implementation for this would return the configured priority of the expression is true.

**Comment by Jetski5822** [comment]
That's a cool idea, I'll open an issue for it later on tomorrow. 

I closed this issue due to the second reply, I don't mind discussing ideas and changes from v1 etc. If we misunderstand or you dislike our answers, I would rather us talk it out than reply with silly comments. Once we worked out where the idea was going a cool idea came out of it, but that idea is not portrayed in the initial open issue.
